File: app/src/main/java/com/pr0gramm/app/ui/fragments/ItemAdminDialog.java
package com.pr0gramm.app.ui.fragments;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.widget.ArrayAdapter;
import android.widget.Checkable;
import android.widget.EditText;
import android.widget.ListView;

import com.f2prateek.dart.InjectExtra;
import com.google.common.primitives.Floats;
import com.jakewharton.rxbinding.widget.RxAdapterView;
import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.R;
import com.pr0gramm.app.feed.FeedItem;
import com.pr0gramm.app.services.AdminService;
import com.pr0gramm.app.ui.DialogBuilder;
import com.pr0gramm.app.ui.base.BaseDialogFragment;

import javax.inject.Inject;

import butterknife.BindView;
import rx.Observable;

import static com.google.common.base.MoreObjects.firstNonNull;
import static com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.defaultOnError;

/**
 */
public class ItemAdminDialog extends BaseDialogFragment {
    private static final String KEY_FEED_ITEM = "ItemAdminDialog__feedItem";

    @Inject
    AdminService adminService;

    @InjectExtra(KEY_FEED_ITEM)
    FeedItem item;

    @BindView(R.id.reason)
    ListView reasonListView;

    @BindView(R.id.custom_reason)
    EditText customReasonText;

    @BindView(R.id.block_user)
    Checkable blockUser;

    @BindView(R.id.block_user_days)
    EditText blockUserForDays;

    @BindView(R.id.notify_user)
    Checkable notifyUser;

    @Override
    protected void injectComponent(ActivityComponent activityComponent) {
        activityComponent.inject(this);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return DialogBuilder.start(getContext())
                .layout(R.layout.admin_delete_item)
                .negative(R.string.cancel, this::dismiss)
                .positive(R.string.delete, di -> onDeleteClicked())
                .noAutoDismiss()
                .build();
    }

    @Override
    protected void onDialogViewCreated() {
        reasonListView.setAdapter(new ArrayAdapter<>(getDialog().getContext(),
                android.R.layout.simple_list_item_1, AdminService.REASONS));

        RxAdapterView.itemClicks(reasonListView).subscribe(index -> {
            customReasonText.setText(AdminService.REASONS.get(index));
        });
    }

    private void onDeleteClicked() {
        String reason = customReasonText.getText().toString().trim();
        boolean notifyUser = this.notifyUser.isChecked();
        boolean ban = blockUser.isChecked();

        if (reason.isEmpty()) {
            return;
        }

        Observable<Void> response = deleteItem(reason, notifyUser, ban);
        response.compose(bindToLifecycleAsync())
                .lift(BusyDialogFragment.busyDialog(this))
                .subscribe(event -> dismiss(), defaultOnError());

    }

    private Observable<Void> deleteItem(String reason, boolean notifyUser, boolean ban) {
        if (ban) {
            float banUserDays = firstNonNull(Floats.tryParse(blockUserForDays.getText().toString()), 1.f);
            return adminService.deleteItem(item, reason, notifyUser, banUserDays);
        } else {
            return adminService.deleteItem(item, reason, notifyUser);
        }
    }

    public static ItemAdminDialog newInstance(FeedItem item) {
        Bundle args = new Bundle();
        args.putParcelable(KEY_FEED_ITEM, item);

        ItemAdminDialog dialog = new ItemAdminDialog();
        dialog.setArguments(args);
        return dialog;
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/fragments/ItemAdminDialog.kt
package com.pr0gramm.app.ui.fragments

import android.app.Dialog
import android.os.Bundle
import android.widget.ArrayAdapter
import android.widget.CheckBox
import android.widget.EditText
import android.widget.ListView
import com.github.salomonbrys.kodein.instance
import com.google.common.base.MoreObjects.firstNonNull
import com.google.common.primitives.Floats
import com.jakewharton.rxbinding.widget.itemClicks
import com.pr0gramm.app.R
import com.pr0gramm.app.feed.FeedItem
import com.pr0gramm.app.services.AdminService
import com.pr0gramm.app.ui.base.BaseDialogFragment
import com.pr0gramm.app.ui.dialog
import com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.Companion.defaultOnError
import com.pr0gramm.app.util.arguments
import rx.Completable
import rx.functions.Action0

/**
 */
class ItemAdminDialog : BaseDialogFragment() {
    private val adminService: AdminService by instance()

    private val reasonListView: ListView by bindView(R.id.reason)
    private val customReasonText: EditText by bindView(R.id.custom_reason)
    private val blockUser: CheckBox by bindView(R.id.block_user)
    private val blockUserForDays: EditText by bindView(R.id.block_user_days)
    private val notifyUser: CheckBox by bindView(R.id.notify_user)

    private val item by lazy { arguments.getParcelable<FeedItem>(KEY_FEED_ITEM) }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return dialog(context) {
            layout(R.layout.admin_delete_item)
            negative(R.string.cancel) { dismiss() }
            positive(R.string.delete) { onDeleteClicked() }
            noAutoDismiss()
        }
    }

    override fun onDialogViewCreated() {
        reasonListView.adapter = ArrayAdapter(dialog.context,
                android.R.layout.simple_list_item_1, AdminService.REASONS)

        reasonListView.itemClicks().subscribe { index ->
            customReasonText.setText(AdminService.REASONS[index])
        }
    }

    private fun onDeleteClicked() {
        val reason = customReasonText.text.toString().trim()
        val notifyUser = this.notifyUser.isChecked
        val ban = blockUser.isChecked

        if (reason.isEmpty()) {
            return
        }

        deleteItem(reason, notifyUser, ban)
                .compose(bindToLifecycleAsync<Any>().forCompletable())
                .withBusyDialog(this)
                .subscribe(Action0 { this.dismiss() }, defaultOnError())

    }

    private fun deleteItem(reason: String, notifyUser: Boolean, ban: Boolean): Completable {
        if (ban) {
            val banUserDays = firstNonNull(Floats.tryParse(blockUserForDays.text.toString()), 1f)
            return adminService.deleteItem(item, reason, notifyUser, banUserDays)
        } else {
            return adminService.deleteItem(item, reason, notifyUser)
        }
    }

    companion object {
        private const val KEY_FEED_ITEM = "feedItem"

        fun newInstance(item: FeedItem) = ItemAdminDialog().arguments {
            putParcelable(KEY_FEED_ITEM, item)
        }
    }
}
